src/CommandLine.cc

```diff
@@ -0,0 +1,261 @@
+
+    bool handle(const char* cmd, int len);
+    void run();
+};
+
+
+CommandLine::CommandLine()
+    : m_impl(new implementation)
+{ }
+
+CommandLine::~CommandLine() = default;
+
+static const char* prompt(EditLine*)
+{
+    return "runos> ";
+}
+
+void CommandLine::init(Loader*, const Config& rootConfig)
+{
+    const auto& config = config_cd(rootConfig, "cli");
+
+    history(m_impl->hist.get(), &m_impl->hev, H_SETSIZE,
+            config_get(config, "history-size", 800));
+
+    const char* argv0 = QCoreApplication::arguments().at(0).toLatin1().data();
+    m_impl->el.reset(el_init(argv0, stdin, stdout, stderr));
+
+    el_set(m_impl->el.get(), EL_PROMPT,
+            &prompt);
+    el_set(m_impl->el.get(), EL_EDITOR,
+            config_get(config, "editor", "emacs").c_str());
+    el_set(m_impl->el.get(), EL_HIST, history,
+            m_impl->hist.get());
+    register_builtins();
+}
+
+void CommandLine::register_builtins() {
+    { // commands
+    registerCommand("commands", [=](auto vm) {
+                std::vector<std::string> commands_name;
+                commands_name.reserve(m_impl->commands.size());
+                for (const auto& p : m_impl->commands) {
+                    commands_name.push_back(p.first);
+                }
+                std::sort(commands_name.begin(), commands_name.end());
+                for (const auto& name : commands_name) {
+                    std::cout << name << std::endl;
+                }
+            }, "List of commands");
+    } // commands
+
+    { // help
+        cli::options::options_description desc;
+        desc.add_options()
+             ("command-name", cli::options::value<std::string>(), "name of command");
+        cli::options::positional_options_description pos;
+        pos.add("command-name", 1)
+           .add("others", -1);
+        auto help_fn =
+            [=](const cli::options::variables_map& vm) {
+                auto arg = vm["command-name"];
+                std::string cmd_name =
+                    arg.empty() ? "help" : arg.as<std::string>();
+                try {
+                    std::cout << cmd_name  << ":" << std::endl << std::endl;
+                    const auto& cmd = m_impl->commands.at(cmd_name);
+                    std::cout << cmd.help;
+                    std::cout << std::endl;
+                    std::cout << cmd.opts;
+                    std::cout << std::endl;
+                }
+                catch (std::out_of_range& oor) {
+                    std::cout << "Unknown command : " << cmd_name << std::endl <<
+                            "Type \"commands\" for show existing commands"
+                            << std::endl;
+                }
+            };
+        registerCommand("help", std::move(desc), std::move(pos), std::move(help_fn),
+                "Help!!! I need somebody\n"
+                "Print help of command\n"
+                "Usage: help <command>\n"
+            );
+    } // help
```

Comment:
What the manner to write such large lambdas? I haven't read this attentively enough, but it seems that this function can be moved out of there.